File: test/gex.test.js
/* Copyright (c) 2013-2022 Richard Rodger, MIT License */

const { describe, test } = require('node:test')
const assert = require('node:assert')

const GexRoot = require('..')
const { Gex } = require('..')

function s(obj) {
  return JSON.stringify(obj)
}

describe('Gex', function () {
  test('happy', () => {
    var ab = Gex('ab')
    assert.equal(ab.on('ab'), 'ab')
    assert.equal(ab.on('a'), null)
    assert.equal(ab.on('b'), null)
    assert.equal(ab.on('ba'), null)
    assert.equal(ab.on('abc'), null)
    assert.equal(ab.on('cab'), null)
    assert.equal(ab.on('cabc'), null)

    var a_b = Gex('a*b')
    assert.equal(a_b.on('acb'), 'acb')
    assert.equal(a_b.on('adb'), 'adb')
    assert.equal(a_b.on('aab'), 'aab')
    assert.equal(a_b.on('abb'), 'abb')

    assert.equal(a_b.on('aaa'), null)
    assert.equal(a_b.on('bbb'), null)
    assert.equal(a_b.on('bca'), null)
    assert.equal(a_b.on('ba'), null)
    assert.equal(a_b.on('ac'), null)
    assert.equal(a_b.on('a'), null)

    var a$b = Gex('a?b')
    assert.equal(a$b.on('acb'), 'acb')
    assert.equal(a$b.on('adb'), 'adb')
    assert.equal(a$b.on('aab'), 'aab')
    assert.equal(a$b.on('abb'), 'abb')

    assert.equal(a$b.on('aaa'), null)
    assert.equal(a$b.on('bbb'), null)
    assert.equal(a$b.on('bca'), null)
    assert.equal(a$b.on('ba'), null)
    assert.equal(a$b.on('ac'), null)
    assert.equal(a$b.on('a'), null)

    assert.equal(a_b.on('accb'), 'accb')
    assert.equal(a_b.on('acccb'), 'acccb')
    assert.equal(a_b.on('aaab'), 'aaab')
    assert.equal(a_b.on('aabb'), 'aabb')
    assert.equal(a_b.on('abbb'), 'abbb')
  })

  test('module-root', () => {
    var ab = GexRoot('ab')
    assert.equal(ab.on('ab'), 'ab')
  })

  test('arrays', () => {
    var a_ = Gex('a*')
    assert.equal(s(a_.on(['ab', 'ac'])), s(['ab', 'ac']))
    assert.equal(s(a_.on(['ab', 'dd', 'ac'])), s(['ab', 'ac']))
    assert.equal(s(a_.on(['ab', 'dd', 'ee'])), s(['ab']))
    assert.equal(s(a_.on(['ff', 'dd', 'ee'])), s([]))
    assert.equal(s(a_.on([])), s([]))
    assert.equal(s(a_.on([null])), s([]))
    assert.equal(s(a_.on(['ab', null, 'dd', undefined, 'ee', NaN])),
      s(['ab']))
  })

  test('objects', () => {
    var foo_ = Gex('foo*')
    assert.equal(s(foo_.on({ foo: 1 })), s({ foo: 1 }))
    assert.equal(s(foo_.on({ foo: 1, doo: 2 })), s({ foo: 1 }))
    assert.equal(s(foo_.on({ foo: 1, doo: 2, food: 3 })),
      s({ foo: 1, food: 3 }))

    var o0 = { food: 3 }
    var o1 = Object.create(o0)
    o1.foo = 1
    o1.doo = 2
    assert.equal(s(foo_.on(o1)), s({ foo: 1 }))
  })

  test('object without prototype', () => {
    var obj = Object.create(null)
    obj.foo = 'bar'
    assert.equal(s({ foo: 'bar' }), s(Gex('foo').on(obj)))
  })

  test('dodgy', () => {
    assert.equal(Gex().on('aaa'), null)
    assert.equal(Gex(null).on('aaa'), null)
    assert.equal(Gex(NaN).on('aaa'), null)
    assert.equal(Gex(undefined).on('aaa'), null)

    var g = Gex('g')
    assert.equal(g.on(), null)
    assert.equal(g.on(null), null)
    assert.equal(g.on(NaN), null)
    assert.equal(g.on(undefined), null)

    assert.equal(g.on(true), null)
    assert.equal(g.on(false), null)
    assert.equal(g.on(new Date()), null)
    assert.equal(g.on(/x/), null)

    assert.equal(g.on(''), null)
  })

  test('escapes', () => {
    var g = Gex('a**b')
    assert.equal(g.toString(), 'Gex[a**b]')
    assert.equal(g.re().toString(), '/^a\\*b$/')
    assert.equal(g.on('a*b'), 'a*b')
    assert.equal(g.on('a**b'), null) // not a literal 'a*b'

    g = Gex('a*?b')
    assert.equal(g.toString(), 'Gex[a*?b]')
    assert.equal(g.re().toString(), '/^a\\?b$/')
    assert.equal(g.on('a?b'), 'a?b')
    assert.equal(g.on('a*?b'), null) // not a literal 'a?b'

    assert.equal(g.esc(''), '')
    assert.equal(g.esc('*'), '**')
    assert.equal(g.esc('?'), '*?')
    assert.equal(g.esc('a*'), 'a**')
    assert.equal(g.esc('a?'), 'a*?')
    assert.equal(g.esc('a*b*c'), 'a**b**c')
    assert.equal(g.esc('a?b?c'), 'a*?b*?c')
  })

  test('newlines', () => {
    var g = Gex('a*b')
    assert.equal('/^a[\\s\\S]*b$/', '' + g.re())
    assert.equal(g.on('a\nb'), 'a\nb')
  })

  test('zero', () => {
    assert.equal(Gex('0').on('0'), '0')
    assert.equal(Gex('0*').on('0'), '0')
    assert.equal(Gex('*0').on('0'), '0')
    assert.equal(Gex('*0*').on('0'), '0')

    assert.equal(Gex(['0']).on('0'), '0')
    assert.equal(Gex(['0*']).on('0'), '0')
    assert.equal(Gex(['*0']).on('0'), '0')
    assert.equal(Gex(['*0*']).on('0'), '0')

    assert.equal(Gex(1).on('1'), '1')
    assert.equal(Gex(100).on('100'), '100')
    assert.equal(Gex(0).on('0'), '0')
  })

  test('multi', () => {
    var g = Gex(['a', 'b'])
    assert.equal(g.on('a'), 'a')
    assert.equal(g.on('b'), 'b')
    assert.equal(s(g.re()), '{"a":{},"b":{}}')

    g = Gex(['a*', 'b'])
    assert.equal(g.on('ax'), 'ax')
    assert.equal(g.on('b'), 'b')
    assert.equal(s(g.re()), '{"a*":{},"b":{}}')

    assert.equal(Gex(['a*', 'b*']).on('bx'), 'bx')
    assert.equal(Gex(['a*', 'b*']).on(['ax', 'zz', 'bx']).toString(), 'ax,bx')
  })

  test('inspect', () => {
    var g = Gex('a*')
    assert.equal(g.toString(), 'Gex[a*]')
    assert.equal(g.inspect(), 'Gex[a*]')

    g = Gex(['a*', '*b'])
    assert.equal(g.inspect(), 'Gex[a*,*b]')
  })

  test('funky', () => {
    assert.equal(Gex('').on('a'), null)
    assert.equal(Gex().on('a'), null)
    assert.equal(Gex(null).on('a'), null)
    assert.equal(Gex(undefined).on('a'), null)
    assert.equal(Gex(NaN).on('a'), null)
    assert.equal(Gex(/a/).on('a'), null)
  })
})

File: go/gex_test.go
// Copyright (c) 2013-2026 Richard Rodger and other contributors, MIT License

package gex

import (
	"math"
	"reflect"
	"testing"
	"time"
)

func expectEqual(t *testing.T, got, want any) {
	t.Helper()
	if !reflect.DeepEqual(got, want) {
		t.Errorf("expected %#v, got %#v", want, got)
	}
}

func TestHappy(t *testing.T) {
	ab := New("ab")
	expectEqual(t, ab.On("ab"), "ab")
	expectEqual(t, ab.On("a"), nil)
	expectEqual(t, ab.On("b"), nil)
	expectEqual(t, ab.On("ba"), nil)
	expectEqual(t, ab.On("abc"), nil)
	expectEqual(t, ab.On("cab"), nil)
	expectEqual(t, ab.On("cabc"), nil)

	aStarB := New("a*b")
	expectEqual(t, aStarB.On("acb"), "acb")
	expectEqual(t, aStarB.On("adb"), "adb")
	expectEqual(t, aStarB.On("aab"), "aab")
	expectEqual(t, aStarB.On("abb"), "abb")

	expectEqual(t, aStarB.On("aaa"), nil)
	expectEqual(t, aStarB.On("bbb"), nil)
	expectEqual(t, aStarB.On("bca"), nil)
	expectEqual(t, aStarB.On("ba"), nil)
	expectEqual(t, aStarB.On("ac"), nil)
	expectEqual(t, aStarB.On("a"), nil)

	aQB := New("a?b")
	expectEqual(t, aQB.On("acb"), "acb")
	expectEqual(t, aQB.On("adb"), "adb")
	expectEqual(t, aQB.On("aab"), "aab")
	expectEqual(t, aQB.On("abb"), "abb")

	expectEqual(t, aQB.On("aaa"), nil)
	expectEqual(t, aQB.On("bbb"), nil)
	expectEqual(t, aQB.On("bca"), nil)
	expectEqual(t, aQB.On("ba"), nil)
	expectEqual(t, aQB.On("ac"), nil)
	expectEqual(t, aQB.On("a"), nil)

	expectEqual(t, aStarB.On("accb"), "accb")
	expectEqual(t, aStarB.On("acccb"), "acccb")
	expectEqual(t, aStarB.On("aaab"), "aaab")
	expectEqual(t, aStarB.On("aabb"), "aabb")
	expectEqual(t, aStarB.On("abbb"), "abbb")
}

func TestArrays(t *testing.T) {
	g := New("a*")
	expectEqual(t, g.On([]string{"ab", "ac"}), []any{"ab", "ac"})
	expectEqual(t, g.On([]string{"ab", "dd", "ac"}), []any{"ab", "ac"})
	expectEqual(t, g.On([]string{"ab", "dd", "ee"}), []any{"ab"})
	expectEqual(t, g.On([]string{"ff", "dd", "ee"}), []any{})
	expectEqual(t, g.On([]string{}), []any{})

	expectEqual(t, g.On([]any{nil}), []any{})
	expectEqual(t, g.On([]any{"ab", nil, "dd", "ee", math.NaN()}), []any{"ab"})
}

func TestObjects(t *testing.T) {
	g := New("foo*")
	expectEqual(t, g.On(map[string]any{"foo": 1}), map[string]any{"foo": 1})
	expectEqual(t, g.On(map[string]any{"foo": 1, "doo": 2}), map[string]any{"foo": 1})
	expectEqual(t, g.On(map[string]any{"foo": 1, "doo": 2, "food": 3}),
		map[string]any{"foo": 1, "food": 3})

	expectEqual(t, g.On(map[string]int{"foo": 1, "doo": 2}), map[string]any{"foo": 1})
}

func TestDodgy(t *testing.T) {
	expectEqual(t, New().On("aaa"), nil)

	g := New("g")
	expectEqual(t, g.On(nil), nil)
	expectEqual(t, g.On(true), nil)
	expectEqual(t, g.On(false), nil)
	expectEqual(t, g.On(time.Now()), nil)
	expectEqual(t, g.On(math.NaN()), nil)
	expectEqual(t, g.On(""), nil)
}

func TestEscapes(t *testing.T) {
	g := New("a**b")
	expectEqual(t, g.String(), "Gex[a**b]")
	expectEqual(t, g.Re().String(), `^a\*b$`)
	expectEqual(t, g.On("a*b"), "a*b")
	expectEqual(t, g.On("a**b"), nil)

	g = New("a*?b")
	expectEqual(t, g.String(), "Gex[a*?b]")
	expectEqual(t, g.Re().String(), `^a\?b$`)
	expectEqual(t, g.On("a?b"), "a?b")
	expectEqual(t, g.On("a*?b"), nil)

	expectEqual(t, g.Esc(""), "")
	expectEqual(t, g.Esc("*"), "**")
	expectEqual(t, g.Esc("?"), "*?")
	expectEqual(t, g.Esc("a*"), "a**")
	expectEqual(t, g.Esc("a?"), "a*?")
	expectEqual(t, g.Esc("a*b*c"), "a**b**c")
	expectEqual(t, g.Esc("a?b?c"), "a*?b*?c")
}

func TestNewlines(t *testing.T) {
	g := New("a*b")
	expectEqual(t, g.Re().String(), `^a[\s\S]*b$`)
	expectEqual(t, g.On("a\nb"), "a\nb")
}

func TestZero(t *testing.T) {
	expectEqual(t, New("0").On("0"), "0")
	expectEqual(t, New("0*").On("0"), "0")
	expectEqual(t, New("*0").On("0"), "0")
	expectEqual(t, New("*0*").On("0"), "0")

	expectEqual(t, New("0").On(0), 0)
	expectEqual(t, New("1").On(1), 1)
	expectEqual(t, New("100").On(100), 100)
}

func TestMulti(t *testing.T) {
	g := New("a", "b")
	expectEqual(t, g.On("a"), "a")
	expectEqual(t, g.On("b"), "b")

	keys := g.Specs()
	expectEqual(t, keys, []string{"a", "b"})

	rmap := g.ReMap()
	if _, ok := rmap["a"]; !ok {
		t.Errorf("ReMap missing 'a'")
	}
	if _, ok := rmap["b"]; !ok {
		t.Errorf("ReMap missing 'b'")
	}
	if g.Re() != nil {
		t.Errorf("Re() should be nil for multi-spec Gex")
	}

	g = New("a*", "b")
	expectEqual(t, g.On("ax"), "ax")
	expectEqual(t, g.On("b"), "b")

	expectEqual(t, New("a*", "b*").On("bx"), "bx")
	expectEqual(t, New("a*", "b*").On([]string{"ax", "zz", "bx"}), []any{"ax", "bx"})
}

func TestInspect(t *testing.T) {
	g := New("a*")
	expectEqual(t, g.String(), "Gex[a*]")
	expectEqual(t, g.Inspect(), "Gex[a*]")

	g = New("a*", "*b")
	expectEqual(t, g.Inspect(), "Gex[a*,*b]")
}

func TestFunky(t *testing.T) {
	expectEqual(t, New("").On("a"), nil)
	expectEqual(t, New().On("a"), nil)
}

func TestMatch(t *testing.T) {
	g := New("a*")
	if !g.Match("abc") {
		t.Errorf("expected 'abc' to match 'a*'")
	}
	if g.Match("xyz") {
		t.Errorf("expected 'xyz' not to match 'a*'")
	}
}

func TestEscPackage(t *testing.T) {
	if Esc("a*b") != "a**b" {
		t.Errorf("package-level Esc broke")
	}
}
